Use pidfd_spawn for faster process spawning when a PidFd is requested

the8472 opened the pull request:
glibc 2.39 added `pidfd_spawnp` and `pidfd_getpid` which makes it possible to get pidfds while staying on the CLONE_VFORK path.

verified that vfork gets used with strace:

```
$  strace -ff -e pidfd_open,clone3,openat,execve,waitid,close ./x test std --no-doc -- pidfd
[...]
[pid 2820532] clone3({flags=CLONE_VM|CLONE_PIDFD|CLONE_VFORK|CLONE_CLEAR_SIGHAND, pidfd=0x7b7f885fec6c, exit_signal=SIGCHLD, stack=0x7b7f88aff000, stack_size=0x9000}strace: Process 2820533 attached
 <unfinished ...>
[pid 2820533] execve("/home/the8472/bin/sleep", ["sleep", "1000"], 0x7ffdd0e268d8 /* 107 vars */) = -1 ENOENT (No such file or directory)
[pid 2820533] execve("/home/the8472/.cargo/bin/sleep", ["sleep", "1000"], 0x7ffdd0e268d8 /* 107 vars */) = -1 ENOENT (No such file or directory)
[pid 2820533] execve("/usr/local/bin/sleep", ["sleep", "1000"], 0x7ffdd0e268d8 /* 107 vars */) = -1 ENOENT (No such file or directory)
[pid 2820533] execve("/usr/bin/sleep", ["sleep", "1000"], 0x7ffdd0e268d8 /* 107 vars */ <unfinished ...>
[pid 2820532] <... clone3 resumed> => {pidfd=[3]}, 88) = 2820533
[pid 2820533] <... execve resumed>)     = 0
[pid 2820532] openat(AT_FDCWD, "/proc/self/fdinfo/3", O_RDONLY|O_CLOEXEC) = 4
[pid 2820532] close(4)                  = 0
```

Tracking issue: #82971

the8472 committed:
use pidfd_spawn for faster process creation when pidfds are requested

the8472 committed:
document the cvt methods

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9625663836/job/26550780239) [(plain)](https://github.com/rust-lang/rust/commit/50c48fdcf3bf1cf2cf79cb20848cfd5245703a74/checks/26550780239/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking addr2line v0.22.0
error[E0425]: cannot find value `SYS_pidfd_open` in crate `libc`
   --> library/std/src/sys/pal/unix/process/process_unix.rs:492:50
    |
492 |                     unsafe { libc::syscall(libc::SYS_pidfd_open, our_pid, 0) } as libc::c_int;

For more information about this error, try `rustc --explain E0425`.
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:01:11
```

</details>

the8472 commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -640,6 +681,46 @@ impl Command {
+                let pid = match cvt(pidfd_getpid.get().unwrap()(pidfd)) {
+                    Ok(pid) => pid,
+                    Err(e) => {
+                        // The child has been spawned and we are holding its pidfd.
+                        // But we cannot obtain its pid even though pidfd_getpid support was verified earlier.
+                        // This might happen if libc can't open procfs because the file descriptor limit has been reached.
+                        libc::close(pidfd);
+                        return Err(Error::new(
+                            e.kind(),
+                            "pidfd_spawnp succeeded but the child's PID could not be obtained",
+                        ));
```
In their unfathomable wisdom the glibc developers have not designed `pidfd_spawnp` to return the pid and pidfd at the same time even though `clone3` does support it, which means we can get into this awkward halfway state.

I'm not sure if this is the best solution for the edge-case. An alternative is to change it to  `Process { pid: Option<NonZero<pid_t>> }`  and delay the error reporting by only panicking if someone tries to call `Child.id()` if the when a pid couldn't be obtained.

workingjubilee replied:
lmao.

workingjubilee replied:
I think for right now this is best because I think assuming calling `.id()` won't panic is pretty reasonable and this will let us find out if we do in fact have to make that change. At a glance, it seems like a bit of a "it almost-never _really_ happens and there's no reasonable response" scenario, in which case failing fast is better?

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -449,17 +449,70 @@ impl Command {
+                static PIDFD_SPAWN_SUPPORTED: AtomicU8 = AtomicU8::new(0);
+                const UNKNOWN: u8 = 0;
+                const YES: u8 = 1;
+                // NO currently forces a fallback to fork/exec. We could be more nuanced here and keep using spawn
+                // if we know pidfd's aren't supported at all and the fallback would be futile.
+                const NO: u8 = 2;
```
Is there a particular reason to not do that now?

the8472 replied:
I wanted to keep things simple, but I can add it too.

the8472 replied:
added

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -449,17 +449,70 @@ impl Command {
+                    }
+                }
+            } else {
+                if self.get_create_pidfd() {
+                    unreachable!("only implemented on linux")
+                }
```
Hmm. This currently returns false for all other platforms?

the8472 replied:
yes, pidfd is a linux feature

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -449,17 +449,70 @@ impl Command {
+                }
+
+                weak! { fn pidfd_getpid(libc::c_int) -> libc::c_int }
+
+                static PIDFD_SPAWN_SUPPORTED: AtomicU8 = AtomicU8::new(0);
+                const UNKNOWN: u8 = 0;
+                const YES: u8 = 1;
+                // NO currently forces a fallback to fork/exec. We could be more nuanced here and keep using spawn
+                // if we know pidfd's aren't supported at all and the fallback would be futile.
+                const NO: u8 = 2;
+
+                if self.get_create_pidfd() {
+                    let flag = PIDFD_SPAWN_SUPPORTED.load(Ordering::Relaxed);
+                    if flag == NO || pidfd_spawnp.get().is_none() || pidfd_getpid.get().is_none() {
+                        return Ok(None);
+                    }
+                    if flag == UNKNOWN {
+                        let mut support = NO;
+                        let our_pid = crate::process::id();
+                        let pidfd =
+                            unsafe { libc::syscall(libc::SYS_pidfd_open, our_pid, 0) } as libc::c_int;
+                        if pidfd >= 0 {
+                            let pid = unsafe { pidfd_getpid.get().unwrap()(pidfd) } as u32;
+                            unsafe { libc::close(pidfd) };
+                            if pid == our_pid {
+                                support = YES
+                            };
+                        }
+                        PIDFD_SPAWN_SUPPORTED.store(support, Ordering::Relaxed);
+                        if support != YES {
+                            return Ok(None);
+                        }
+                    }
+                }
+            } else {
+                if self.get_create_pidfd() {
+                    unreachable!("only implemented on linux")
+                }
+            }
+        }
```
Is it possible to outline this code into a support function and just check for and propagate the `Ok(None)` return?

the8472 replied:
The weak symbols are needed later in the same function

the8472 replied:
I'll try moving things into a module

workingjubilee replied:
Oh hm.

the8472 replied:
Tried moving it to a module but `weak!` makes it really unergonomic  since it generates `let`s. Turning things into items requires repeating the function signature several times.

workingjubilee replied:
ugh, annoying.

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -640,6 +690,47 @@ impl Command {
             let spawn_fn = retrying_libc_posix_spawnp;
 
+            #[cfg(target_os = "linux")]
+            if self.get_create_pidfd() {
+                let mut pidfd: libc::c_int = -1;
+                let spawn_res = pidfd_spawnp.get().unwrap()(
+                    &mut pidfd,
+                    self.get_program_cstr().as_ptr(),
+                    file_actions.0.as_ptr(),
+                    attrs.0.as_ptr(),
+                    self.get_argv().as_ptr() as *const _,
+                    envp as *const _,
+                );
+
+                let spawn_res = cvt_nz(spawn_res);
+                if let Err(ref e) = spawn_res
+                    && e.raw_os_error() == Some(libc::ENOSYS)
+                {
+                    PIDFD_SPAWN_SUPPORTED.store(NO, Ordering::Relaxed);
+                    return Ok(None);
+                }
+                spawn_res?;
+
+                let pid = match cvt(pidfd_getpid.get().unwrap()(pidfd)) {
+                    Ok(pid) => pid,
+                    Err(e) => {
+                        // The child has been spawned and we are holding its pidfd.
+                        // But we cannot obtain its pid even though pidfd_getpid support was verified earlier.
+                        // This might happen if libc can't open procfs because the file descriptor limit has been reached.
+                        libc::close(pidfd);
+                        return Err(Error::new(
+                            e.kind(),
+                            "pidfd_spawnp succeeded but the child's PID could not be obtained",
+                        ));
+                    }
+                };
+
+                return Ok(Some(Process::new(pid, pidfd)));
+            }
```
Same question re: outlining this into a support function.

It seems like the answer is that it would be more of a hassle, though.

workingjubilee commented on library/std/src/sys/pal/unix/mod.rs:
```diff
@@ -307,10 +307,13 @@ macro_rules! impl_is_minus_one {
 
 impl_is_minus_one! { i8 i16 i32 i64 isize }
 
+/// Convert native return values to Result using the *-1 means error is in `errno`*  convention.
+/// Non-error values are `Ok`-wrapped.
```
Oh.

Thank you for documenting these, I forget what they mean every time.

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -640,6 +690,47 @@ impl Command {
+                return Ok(Some(Process::new(pid, pidfd)));
+            }
+
+            // Safety: -1 indicates we don't have a pidfd.
+            let mut p = Process::new(0, -1);
```
...Sorry, can you add the Safety condition to the `Process::new` function as well? It's not immediately clear why it is `unsafe` to begin with... It's fine to add "this list is non-exhaustive", too.

the8472 replied:
done

the8472 committed:
document safety properties of the internal Process::new constructor

the8472 committed:
more fine-grained feature-detection for pidfd spawning

we now distinguish between pidfd_spawn support, pidfd-via-fork/exec and not-supported

the8472 committed:
Check that we get somewhat sane PIDs when spawning with pidfds

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9653197332/job/26624996988) [(plain)](https://github.com/rust-lang/rust/commit/2a85c05d45676dd75bb5427443d9ef8e2de20846/checks/26624996988/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9653385479/job/26625546867) [(plain)](https://github.com/rust-lang/rust/commit/ed60a16d03104f442f4d5f08ce12934fd9cf44d5/checks/26625546867/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting std v0.0.0 (/checkout/library/std)
error: unresolved link to `std::io`
   --> library/std/src/sys/pal/unix/process/process_unix.rs:897:23
    |
897 |     /// [I/O Safety]: std::io#io-safety
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

```

</details>

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -640,6 +702,47 @@ impl Command {
+            #[cfg(target_os = "linux")]
+            if self.get_create_pidfd() && PIDFD_SUPPORTED.load(Ordering::Relaxed) == SPAWN {
+                let mut pidfd: libc::c_int = -1;
+                let spawn_res = pidfd_spawnp.get().unwrap()(
+                    &mut pidfd,
+                    self.get_program_cstr().as_ptr(),
+                    file_actions.0.as_ptr(),
+                    attrs.0.as_ptr(),
+                    self.get_argv().as_ptr() as *const _,
+                    envp as *const _,
+                );
```
Ah, I see why they didn't add another argument to `pidfd_spawnp`: this way it is "exactly the same" arg/ret as `posix_spawnp` instead of adding another one. They are 1:1 except for "you get a pidfd, not a pid", right?

workingjubilee commented on library/std/src/sys/pal/unix/process/process_unix.rs:
```diff
@@ -449,17 +449,82 @@ impl Command {
+                            }
+                            _ => {}
+                        }
+                        PIDFD_SUPPORTED.store(support, Ordering::Relaxed);
```
Relaxed store because we don't care if we accidentally probe twice in a hypothetical "multiple threads spawn processes with similar Commands" cases, right?

workingjubilee reviewed (approved):
This looks fine to me.

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit ec0c755704bba1b6c4faa0b10aa0d886cdfa309e has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee ec0c755704bba1b6c4faa0b10aa0d886cdfa309e -->
<!-- homu: {"type":"Approved","sha":"ec0c755704bba1b6c4faa0b10aa0d886cdfa309e","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
